feat: Implement a SQL "None" dialect to skip processing EXEC SQL blocks

mm-broadcom opened the pull request:
Created a new SQL "dialect" that allows the user to skip processing any and all text within a `EXEC SQL` and `END-EXEC` block. This has to be an explicitly set option within the extension's settings.

## How Has This Been Tested?
Created two unit tests for the "dialect". One with valid and another with "invalid" code within an EXEC block. Also ran locally to check for visual oddities along with running the myriad of existing tests.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Add SKIP_SQL as a selectable dialect

Added SKIP_SQL as a selectable dialect. This will allow any and all code within EXEC SQL and END-EXEC blocks without any further processing.

mm-broadcom committed:
Adjust ANTLR rule, account for lack of implicit copybooks

mm-broadcom committed:
Minimize code duplication for DB2/Skip-SQL

mm-broadcom committed:
Change SKIP SQL's description

Changed the description to better represent what the user is getting. This does have some minor spell checking and other aspects but does not contain the deeper aspects of the grammar.

mm-broadcom committed:
Skip errors when in SKIP_SQL

mm-broadcom committed:
Remove unnecessary changes to ExecParser

mm-broadcom committed:
Add dialect check to visitSqlCode for SKIP_SQL

mm-broadcom committed:
Removed missed line from ExecParser

mm-broadcom committed:
Rephrase description

mm-broadcom committed:
Ran formatter

mm-broadcom committed:
Restore ExecParser to original state

slavek-kucera reviewed (changes requested):
![image](https://github.com/user-attachments/assets/8c1f1487-5264-4676-a9e1-e07183f5414d)

mm-broadcom committed:
Change SKIP_SQL to a boolean option

Change the SKIP_SQL "dialect" to a boolean option. This is to avoid issues with implicit copybooks when trying to skip the majority of the SQL processing.

Also adjusted constructors within tests to account for the new parameter.

mm-broadcom committed:
Removed alternate start rule from grammar

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/copybook/SQLBackend.java:
```diff
@@ -21,5 +21,6 @@
 public enum SQLBackend {
   DB2_SERVER,
   DATACOM_SERVER,
+  SKIP_SQL,
```
I think this can be reverted now.
```suggestion
```

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlDialect.java:
```diff
@@ -71,6 +72,7 @@ public Map<String, String> getKeywords() {
 
   @Override
   public ResultWithErrors<DialectOutcome> processText(DialectProcessingContext context) {
+    boolean isSqlProcessingEnabled = getSqlProcessingEnabled(context);
     Db2SqlVisitor db2SqlVisitor = new Db2SqlVisitor(context, messageService, copybookService);
```
```suggestion
    boolean isSqlProcessingEnabled = getSqlProcessingEnabled(context);
    Db2SqlVisitor db2SqlVisitor = new Db2SqlVisitor(context, messageService, copybookService, isSqlProcessingEnabled);
```

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlDialect.java:
```diff
@@ -89,7 +92,7 @@ public ResultWithErrors<DialectOutcome> processText(DialectProcessingContext con
 
     // Add error encountered while visiting the parser. To be reported to COBOL LS engine.
-    parseError.addAll(db2SqlVisitor.getErrors());
+    if (isSqlProcessingEnabled) parseError.addAll(db2SqlVisitor.getErrors());
```
Don't produce the errors in the first place.
```suggestion
    parseError.addAll(db2SqlVisitor.getErrors());
```

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/AnalysisConfig.java:
```diff
@@ -34,6 +34,7 @@ public class AnalysisConfig {
   List<String> dialects;
   boolean isCicsTranslatorEnabled;
   boolean collectAstChanges;
+  boolean isSqlProcessingEnabled;
```
If this is the direction we'll end up taking, we'll need to clean up the booleans and turn them into enums.

slavek-kucera replied:
3 booleans is too much... Create an enum and default the variable to the enabled state - that way you should be able to get rid of bunch of the `true,` values added all over..

mm-broadcom replied:
Done in latest commit

mm-broadcom committed:
Remove SKIP_SQL mentions

mm-broadcom committed:
Implemented requested changes

mm-broadcom committed:
Group SQL Settings

- Added `sql.` prefix before the TARGET_BACKEND and PROCESSING_ENABLED settings such that they are grouped together in the settings.
- Renamed mentions of `target-sql-backend`` to `sql.target-backend`.

mm-broadcom committed:
Improve Enable SQL Processing description

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -488,7 +488,7 @@
           "description": "The logging level for COBOL LS backend"
         },
-        "cobol-lsp.target-sql-backend": {
+        "cobol-lsp.sql.target-backend": {
```
This would be a breaking change.

mm-broadcom replied:
There is no other way to group settings within an extension. VS Code puts all settings in alphabetical order when you view the GUI rather than the raw settings JSON file. 

At _most_ we could try to keep them together by having the boolean for enabling/disabling SQL processing named something along the lines of `target-sql-processing-enabled`.

mm-broadcom committed:
Restore target-sql-backend name, rename processing setting

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlDialect.java:
```diff
@@ -149,10 +151,19 @@ private Db2SqlParser.StartRuleContext parseDB2(
 
     Db2SqlParser.StartRuleContext result = parser.startRule();
-    errors.addAll(listener.getErrors());
+    if (isSqlProcessingEnabled) errors.addAll(listener.getErrors());
```
I think this will also suppress diagnostics like missing END-EXEC.

mm-broadcom replied:
Added code to pass along the END-EXEC error if it is present and SQL processing is off while suppressing the other ones from the SQL grammar.

Also created a test to check this functionality.

mm-broadcom committed:
Provide missing END-EXEC error if present and SQL processing is disabled

mm-broadcom committed:
Ran formatter

mm-broadcom committed:
Removed unnecessary comment

slavek-kucera commented on server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlDialect.java:
```diff
@@ -149,10 +151,29 @@ private Db2SqlParser.StartRuleContext parseDB2(
 
     Db2SqlParser.StartRuleContext result = parser.startRule();
-    errors.addAll(listener.getErrors());
+    if (isSqlProcessingEnabled) {
+      errors.addAll(listener.getErrors());
+    } else {
+      String endExecMessage = messageService.getMessage("db2Parser.missingEndExec");
+      List<SyntaxError> currentErrors = listener.getErrors();
+      Optional<SyntaxError> missingEndExecErr =
+          currentErrors.stream()
+              .filter(err -> endExecMessage.equalsIgnoreCase(err.getSuggestion()))
+              .findFirst();
+      missingEndExecErr.ifPresent(errors::add);
+    }
```
```suggestion
    errors.addAll(listener.getErrors());
```

mm-broadcom committed:
Update server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlDialect.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Removed unused parameter

mm-broadcom committed:
Ran formatter

slavek-kucera commented on .vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Remove the file.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/package.json:
```diff
@@ -503,6 +503,11 @@
+        "cobol-lsp.target-sql-backend-enable-processing": {
+          "type": "boolean",
+          "default": "true",
+          "description": "Enable SQL Processing - If disabled, will still use the target backend to determine implicit copybooks to avoid errors"
```
@zeibura Could you review the description?

zeibura replied:
Is "if disabled, will still use the target backend to determine implicit copybooks to avoid errors" necessary info for the user? It seems like a description of a piece of code which has no impact on user experience. I would remove that part unless there's any reason to keep it.

mm-broadcom replied:
Wanted to let the user know that even though error reporting is disabled, there are still some SQL things going on in the background. Can rephrase or remove if need be.

slavek-kucera replied:
what about 
```
Enables syntactic validation of SQL statements inside EXEC SQL blocks
```
?

mm-broadcom replied:
That works for me. @zeibura What are your thoughts?

zeibura replied:
Fine for now, I'll review this again after August 15.

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/file/WorkspaceFileService.java:
```diff
@@ -162,6 +162,9 @@ public String readImplicitCode(String name) {
     try (InputStream inputStream = ImplicitCodeUtils.class.getResourceAsStream(uri)) {
+      if (inputStream == null) {
+        return "";
+      }
```
What is the condition that causes that?

mm-broadcom replied:
Ran into this case back when "None" was its own dialect. Will test without it in now that "none" is a checkbox.

mm-broadcom replied:
Ran all tests locally with that code commented out, it passed everything so I've fully removed it in the latest commit.

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/PredefinedCopybooks.java:
```diff
@@ -51,7 +51,14 @@ public enum Copybook {
+            return "SQLCA_DATACOM";
+          default:
+            return "SQLCA_DB2";
+        }
```
This can be reverted.
```suggestion
        return backend == SQLBackend.DATACOM_SERVER ? "SQLCA_DATACOM" : "SQLCA_DB2";
```

slavek-kucera commented on server/engine/src/main/antlr4/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlExecParser.g4:
```diff
@@ -1651,4 +1651,4 @@ kmg_blob_parameter: INTEGERLITERAL dbs_k_m_g_identifier? | dbs_sql_identifier;
 dbs_maxPartition: INTEGERLITERAL;
 dbs_minus_one: MINUSCHAR INTEGERLITERAL;
-/////
+/////
```
revert

mm-broadcom committed:
Update server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/PredefinedCopybooks.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Convert booleans to an enum in AnalysisConfig

mm-broadcom committed:
Removed check if inputStream was null

slavek-kucera committed:
Revert "Convert booleans to an enum in AnalysisConfig"

This reverts commit 39be91906bfae6e4c4e11c01789e9d75a0710aaf.

slavek-kucera committed:
sqlprocessing enum

slavek-kucera committed:
rename

slavek-kucera committed:
remove vscode settings.json

mm-broadcom committed:
Update description

slavek-kucera committed:
...

slavek-kucera committed:
revert irrelevant change
